realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/sync/MutableSubscriptionTests.kt

```diff
@@ -0,0 +1,299 @@
+        }
+    }
+
+    @Test
+    fun removeSubscription_unManagedthrows() {
+        realm.subscriptions.update { mutableSubs ->
+            val managedSub = mutableSubs.add(Subscription.create("sub", realm.where<SyncColor>()))
+            assertFailsWith<IllegalArgumentException> {
+                mutableSubs.remove(Subscription.create("sub", realm.where<SyncColor>()))
+            }
+        }
+    }
+
+    @Test
+    fun removeAllStringTyped() {
+        var updatedSubs = realm.subscriptions.update { mutableSubs ->
+            mutableSubs.addOrUpdate(Subscription.create(realm.where<SyncColor>()))
+        }
+        assertEquals(1, updatedSubs.size())
+        updatedSubs = updatedSubs.update { mutableSubs ->
+            assertTrue(mutableSubs.removeAll("SyncColor"))
+            assertEquals(0, mutableSubs.size());
+        }
+        assertEquals(0, updatedSubs.size());
+    }
+
+    @Test
+    fun removeAllStringTyped_fails() {
+        // Not part of schema
+        realm.subscriptions.update { mutableSubs ->
+            assertFalse(mutableSubs.removeAll("DontExists"))
+        }
+
+        // part of schema
+        realm.subscriptions.update { mutableSubs ->
+            assertFalse(mutableSubs.removeAll("SyncColor"))
+        }
+    }
+
+    @Test
+    fun removeAllClassTyped() {
+        var updatedSubs = realm.subscriptions.update { mutableSubs ->
+            mutableSubs.addOrUpdate(Subscription.create(realm.where<SyncColor>()))
+        }
+        assertEquals(1, updatedSubs.size())
+        updatedSubs = updatedSubs.update { mutableSubs ->
+            assertTrue(mutableSubs.removeAll(SyncColor::class.java))
+            assertEquals(0, mutableSubs.size());
+        }
+        assertEquals(0, updatedSubs.size());
+    }
+
+    @Test
+    fun removeAllClassTyped_fails() {
+        // Not part of schema
+        realm.subscriptions.update { mutableSubs ->
+            assertFailsWith<IllegalArgumentException> {
+                mutableSubs.removeAll(SyncDog::class.java)
+            }
+        }
+
+        // part of schema
+        realm.subscriptions.update { mutableSubs ->
+            assertFalse(mutableSubs.removeAll(SyncColor::class.java))
+        }
+    }
+
+    @Test
+    fun removeAll() {
+        var updatedSubs = realm.subscriptions.update { mutableSubs ->
+            mutableSubs.addOrUpdate(Subscription.create("test", realm.where<SyncColor>()))
+        }
+        assertEquals(1, updatedSubs.size())
+        updatedSubs = updatedSubs.update { mutableSubs ->
+            assertTrue(mutableSubs.removeAll())
+            assertEquals(0, mutableSubs.size());
+        }
+        assertEquals(0, updatedSubs.size());
+    }
+
+    @Test
+    fun removeAll_fails() {
+        realm.subscriptions.update { mutableSubs ->
+            assertFalse(mutableSubs.removeAll())
+        }
+    }
+
+    @Test
+    fun waitForSynchronizationAfterInsert() {
+        var updatedSubs = realm.subscriptions.update { mutableSubs ->
```

Comment:
Shall we validate that `waitForSynchronization` is a noop if the state is already `COMPLETE`?